ddtrace/internal/datadog/profiling/stack_v2/src/stack_renderer.cpp

```diff
@@ -33,14 +33,56 @@ StackRenderer::render_thread_begin(PyThreadState* tstate,
     // clock_gettime(CLOCK_MONOTONIC) on linux and mach_absolute_time() on macOS.
     // This is not the same as std::chrono::steady_clock, which is backed by clock_gettime(CLOCK_MONOTONIC_RAW)
     // (although this is underspecified in the standard)
+    int64_t now_ns = 0;
     timespec ts;
     if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
-        auto now_ns = static_cast<int64_t>(ts.tv_sec) * 1'000'000'000LL + static_cast<int64_t>(ts.tv_nsec);
+        now_ns = static_cast<int64_t>(ts.tv_sec) * 1'000'000'000LL + static_cast<int64_t>(ts.tv_nsec);
         ddup_push_monotonic_ns(sample, now_ns);
     }
 
+    // Save the thread information in case we observe a task on the thread
+    thread_state.id = thread_id;
+    thread_state.native_id = native_id;
+    thread_state.name = std::string(name);
+    thread_state.now_time_ns = now_ns;
+    thread_state.wall_time_ns = 1000LL * wall_time_us;
+    thread_state.cpu_time_ns = 0; // Walltime samples are guaranteed, but CPU times are not. Initialize to 0
+                                  // since we don't know if we'll get a CPU time here.
+
+    // Finalize the thread information we have
     ddup_push_threadinfo(sample, static_cast<int64_t>(thread_id), static_cast<int64_t>(native_id), name);
-    ddup_push_walltime(sample, 1000LL * wall_time_us, 1);
+    ddup_push_walltime(sample, thread_state.wall_time_ns, 1);
+}
+
+void
+StackRenderer::render_task_begin(std::string_view name)
```

Comment:
Dumb question, but how was this done before you added `render_task_begin`? Or is the concept of a `task` the new addition that was added in echion?

Reply:
> how was this done before

It was broken!

Task comprehension was done correctly within echion itself, but the rendering system I added (to echion) tried to conflate the concept of a task with the concept of a thread.  I had done it this way because the Profiling backend actually makes no distinction between the two concepts, except that they have different tags (and thus, different UI filters).

Unfortunately, that approach didn't do the state management correctly (tasks in Python live on a thread, and that thread is backed by a native thread, and we need to propagate that information), so I had to fix it upstream, and 

This PR mostly fixes dd-trace-py's consumption of the echion rendering API.